### PR TITLE
Initial Implementation of New Template Workflow

### PR DESCRIPTION
**Pull request by Dtphelan1**
This encompasses an initial implementation of our new template workflow. Additions include: 
- The addition of a TemplateSelectionView that gets displayed when selecting a resume; 
- The ability to search/filter on available templates in a similar fashion to shortcut-search;
- The transition of the TemplateViewMode in the ContextTray to SnippetViewMode;
- A cancel button to short-circuit the template selection
- Added a test to confirm that the new-note button opens the template selection screen, updating dependent tests as needed.

Refactors along the way include: 
- Modifications to several functions involved in the: 
    - The opening of new notes;
    - The selection of a template;
    - The reverting of a template.
- Improvements to the styling and the styling-logic of the ContextTray buttons;
- Separation of the SearchBar used in shortcut-search and template-search into its own component.
- Fixed the close-note button styling for IE (had to fix the Cancel button as well; came for free)
- Organization of the .gitignore css files by alphabetical order - makes the removal/addition of old/new css files easier (Spent more time than I'd care to admit rereading this list trying and failing to find a file I was looking to remove - this saved me time doing that later in my task).
- Improved some old tests that used WrapperElement.at(0)props().onClick() instead of simulating a click via enzymes WrapperElement.simulate('click') functionality. 

_Pull requests into Flux require the following. Submitter should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable (**N/A**) and :white_check_mark:. Reviewers should complete every item on the bulleted list when reviewing._

## Submitter:

**Running the Application**

- [x] Manually tested in Chrome (primary browser for testing)
- [x] Manually tested in IE/Safari (verify app loads and basic feature testing)
- [x] Manually tested in Chrome in iPad mode (at minimum, viewing patient record)

**Task**

- [x] Branch implements task as expected
- [x] Key application features are updated as needed for task and tested (see: [Key Application Features](https://github.com/FluxNotes/flux/wiki/Key-Application-Features))
- [x] Task specific definition of done (as described in the Jira task) is met, if applicable.

**Documentation**

- [x] This pull request describes why these changes were made
- [x] Recognizes any potential shortcomings/bugs in the description 
- [x] Pre release script is updated on the [Pre Release Testing Script wiki page](https://github.com/FluxNotes/flux/wiki/Pre-Release-Testing-Script) (add/remove single features to test, do not run through full script)
- [x] Update the list of key features with new features on the [Key Application Features wiki page](https://github.com/FluxNotes/flux/wiki/Key-Application-Features)
- [x] Document any new APIs/extension points
- [x] Additional technical components and libraries are documented and added to [wiki](https://github.com/FluxNotes/flux/wiki/About-Flux-Notes) as needed

**Code Quality**

- [ ] Code style guide is followed (see: [Code Style Guide](https://github.com/FluxNotes/flux/wiki/JavaScript-and-HTML-Code-Style-Guide))
- [ ] Code is commented

**Tests**

- [ ] Existing tests passed
- [ ] Added backend tests for new functionality added

## Reviewers:

- Check that code is maintainable and reusable. Code reuses existing code and infrastructure where appropriate.
- Check the PR and code accomplishes the task's purpose.
- Check that new tests appropriately test the new code, including edge cases. Other existing tests pass.
- Verify Pull Request checklist is correctly completed by submitter.
- Try to break the code

### CODE REVIEW

**Comment by Dtphelan1** [comment]
[DP1/2] - There's a known issue with min-height on small screens and the positioning of the cancel button I'm working on.

**Comment by Dtphelan1** [comment]
Should have addressed AP1/2

**Comment by jafeltra** [comment]
This looks really good! The new flow is really cool and definitely a lot more clear. A few small things I noticed:

[JA1]: Opening a source note from the TDP while in the middle of inserting a template prints out a console.log and causes the following error: NotesPanel.jsx:228 Tried to remove a note that is signed.
[JA2] Using the app search and hovering over a result that is an existing note from the list causes a setState warning when you are in the middle of choosing a template through the new flow.

File: src/context/ContextTray.jsx [hunk 1]
```diff
@@ -1,37 +1,37 @@
...
         };
     }
 
-    componentWillReceiveProps = (nextProps) => {
-        // If show template view is true, set the view to template view, otherwise check if current view is template view. If the view is currently a template go to shortcut view, else leave the view as is
-        // (ex. if in placeholder view, stay in placeholder view)
-        const viewMode = nextProps.showTemplateView ? this.TEMPLATE_VIEW : (this.state.viewMode === this.TEMPLATE_VIEW ? this.SHORTCUT_VIEW : this.state.viewMode);
-        this.setState ({ viewMode });
-    }
+    // componentWillReceiveProps = (nextProps) => {
```
Comment (jafeltra) [inline src/context/ContextTray.jsx]: Do we still need this old method? Is there information in here we want to keep around?

**Comment by sgober** [comment]
[SO1] is there a reason we don't alphabetize the templates?  Seems that if we have a search option they should be ordered in some consistent way

Related to JA1, if you try and open a source note when you have the template selection view up, nothing happens until you close the note and get the error @jafeltra mentioned.  In addition if you've selected a template but have not selected a condition, the error is thrown immediately and the notes editor closes.

also needs a rebase.  In my branch, I moved `CompassAppDashboard.css` to `.scss` so there might need to be some tweaking there.

**Comment by Dtphelan1** [comment]
Rebased!

Regarding SO1, the search results are ordered by relevance -- the first element is the best match, and so on down the line. When there is no search string, there's no meaningful ordering imposed on the data -- it's just the order in which they are listed in the original array. I could sort them in order of name if desired.

**Comment by Dtphelan1** [comment]
Should Address JA1 but at the cost of adding a new bug where empty new notes get created and crowd the in-progress note section of the note-assistant. After talking with Julia and Matt it doesn't feel like fixing this is of terribly high priority, especially since some of the corner-case logic surrounding this bug will have to change when we incorporate incomplete shortcuts into our approach and remove the pick-list-options UI accordingly.

**Comment by gregquinn2001** [comment]
I would like to discuss JA1 @Dtphelan1

**Comment by Dtphelan1** [comment]
After talking with greg we're going to roll back two commits, capture JA1 in the 13 pointer JIRA and make sure we test that as we complete that task. @jafeltra  or @mgramigna if you could roll that back it would be great.

**Comment by jafeltra** [comment]
I talked to @gregquinn2001 about rolling back the latest changes to this branch. Since we don't want to put any additional effort into changing something that will change again with the 13 point task, we are going to leave the last two commits addressing JA1. We will still test that this case works correctly during the 13 point task work.

**Comment by Dtphelan1** [comment]
Woo hoo from Morocco!   
  

  
  

  
  
>

**Pull request merged**